lib/OpenQA/Worker/Cache.pm

```diff
@@ -35,173 +37,325 @@ our (@ISA, @EXPORT);
-                    local $| = 1;
                     # Don't spam the webui, update only every 5 seconds
                     if (time - $last_updated > 5) {
                         update_setup_status;
+                        toggle_asset_lock($asset, 1);
                         $last_updated = time;
                         if ($progress < $current) {
                             $progress = $current;
-                            print $log "Downloading $asset :", $size == $len ? 100 : $progress, "\n";
+                            print $log "CACHE: Downloading $asset :", $size == $len ? 100 : $progress . "\n";
                         }
                     }
                 });
         });
 
     $tx = $ua->start($tx);
-    if (!$tx->res->is_success) {
-        printf $log "CACHE: download of %s failed with %d: %s\n", basename($asset), $tx->res->code, $tx->res->message;
-        return undef;
+
+    if ($tx->res->code == 304) {
+        if (toggle_asset_lock($asset, 0)) {
+            print "CACHE: Content has not changed, not downloading the $asset but updating last use\n";
+        }
+        else {
+            print "CACHE: Abnormal situation, bailing out\n";
+            $asset = undef;
+        }
+    }
+    elsif ($tx->res->is_success) {
+        $etag = $headers->etag;
+        unlink($asset);
+        $asset = $tx->res->content->asset->move_to($asset)->path;
+        my $size = (stat $asset)[7];
+        if ($size == $headers->content_length) {
+            check_limits($size);
+            update_asset($asset, $etag, $size);
+            print $log "CACHE: Asset download sucessful to $asset, Cache size is: $cache_real_size\n";
+        }
+        else {
+            print $log "CACHE: Size of $asset differs, Expected: "
+              . $headers->content_length
+              . " / Downloaded: "
+              . $size . "\n";
+            $asset = undef;
+        }
     }
     else {
-        print $log "CACHE: " . basename($asset) . " download sucessful\n";
-        $asset = $tx->res->content->asset->move_to($asset);
+        print $log "CACHE: Download of $asset failed with: " . $tx->res->error->{message} . "\n";
+        purge_asset($asset);
+        $asset = undef;
     }
-    close($log);
+
     return $asset;
 }
 
 sub get_asset {
     my ($job, $asset_type, $asset) = @_;
     my $type;
-
-    $asset = catdir($location, $asset);
+    my $result;
+    my $ret;
+    $asset = catdir($location, basename($asset));
 
     while () {
-        read_db();
-        log_debug "CACHE: Aquiring lock";
-        open(my $asset_fd, ">", $asset . ".lock");
 
-        if (!flock($asset_fd, LOCK_EX | LOCK_NB)) {
+        log_debug "CACHE: Aquiring lock for $asset in the database";
+        $result = try_lock_asset($asset);
+        if (!$result) {
             update_setup_status;
-            log_debug("CACHE: Asked to wait for lock, sleeping 10 secs");
-            sleep 10;
+            log_debug "CACHE: wait 5 seconds for the lock.";
```

Comment:
s/wait/waiting/ so that it does not sound like a recommendation to the reader of the log file